client/jetpack-cloud/controller.ts

```diff
@@ -0,0 +1,217 @@
+				{
+					...queryParams,
+					site: undefined,
+				},
+				`${ pathname }/${ siteQuery }`
+			)
+		);
+	}
+};
+
+/**
+ * Fetch site data.
+ *
+ * @param {PageJS.Context} context Route context
+ * @param {number|string} siteId Site id
+ * @returns {Promise} Promise that resolves with the site id and slug
+ */
+const fetchSite = (
+	context: PageJS.Context,
+	siteId: number | string
+): Promise< { id: number | undefined; slug: string | undefined } > => {
+	const { getState, dispatch } = context.store;
+
+	return dispatch( requestSite( siteId ) ).then( () => ( {
+		id: getSiteId( getState(), siteId ),
+		slug: getSiteSlug( getState(), siteId ),
+	} ) );
+};
+
+/**
+ * Store site id in state, and add site to the list of most recent sites.
+ *
+ * @param {PageJS.Context} context Route context
+ * @param {number|string} siteId Site id
+ */
+const selectSite = ( context: PageJS.Context, siteId: number ): void => {
+	const { dispatch } = context.store;
+
+	dispatch( setSelectedSiteId( siteId ) );
+	updateRecentSitesPreferences( context );
+};
+
+/**
+ * Handle case when path contains no site slug. If the current user has only
+ * one site, we try to use this site for context. Otherwise, the user must
+ * select one of their sites.
+ *
+ * @param {PageJS.Context} context Route context
+ * @param {Function} next Next middleware function
+ */
+const siteSelectionWithoutFragment = ( context: PageJS.Context, next: () => void ): void => {
+	const { getState, dispatch } = context.store;
+	const currentUser = getCurrentUser( getState() ) as UserData;
+	const hasOneSite = currentUser?.visible_site_count === 1;
+
+	// User has only one site
+	if ( hasOneSite ) {
+		const primarySiteId = getPrimarySiteId( getState() ) as NonNullable< number >;
+		const primarySiteSlug = getSiteSlug( getState(), primarySiteId );
+
+		// Site is already available in state
+		if ( primarySiteSlug ) {
+			// Redirect to the single site context
+			redirectToPrimary( context, primarySiteSlug );
+		} else {
+			fetchSite( context, primarySiteId ).then( ( { slug } ) => {
+				if ( slug ) {
+					// Redirect to the single site context
+					redirectToPrimary( context, slug );
+				} else {
+					// Display error notification
+					showMissingPrimaryError( currentUser, dispatch );
+				}
+			} );
+		}
+	} else {
+		// Show all sites
+		dispatch( setAllSitesSelected() );
+		next();
+	}
+};
+
+/**
+ * Select site when path contains a site slug.
+ *
+ * @param {string} siteFragment Parsed site slug
+ * @param {PageJS.Context} context Route context
+ * @param {Function} next Next middleware function
+ */
+const siteSelectionWithFragment = (
+	siteFragment: string,
+	context: PageJS.Context,
+	next: () => void
+): void => {
+	const { getState } = context.store;
+	const siteId = getSiteId( getState(), siteFragment );
+
+	// Site is already available in state
+	if ( siteId ) {
+		selectSite( context, siteId );
+		next();
+	} else {
+		fetchSite( context, siteFragment ).then( ( { id } ) => {
```

Comment:
Would it make sense to write this as an async/await function?

Reply:
Do you mean like `const site = await fetchSite( context, siteFragment );`? Do you feel it'd be easier to read?

Reply:
Yes, or more like `const { id } = await fetchSite( context, siteFragment );`. I feel putting logic in the `then` callback can get confusing, especially if this code block grows.